### DISCUSSION TITLE
Dependency discovery for Terraform using the terraform lock file (.terraform.lock.hcl)

### DISCUSSION THREAD

**Discussion by candrews** [aquasecurity/trivy#7566] [Ideas]
### Description

Terraform supports dependency discovery for many languages / build systems already as documented at  https://aquasecurity.github.io/trivy/v0.40/docs/vulnerability/detection/language/

I'd like Trivy to  be able to do the same for Terraform via Terraform's lock file, `.terraform.lock.hcl`, documented at https://developer.hashicorp.com/terraform/language/files/dependency-lock

My goal is to run Trivy to generate an SBOM which includes dependencies expressed in `.terraform.lock.hcl`.

For example, from a directory containing `.terraform.lock.hcl`, I'd like to run:
`docker run -it -v "$(pwd)":/app:Z aquasec/trivy:latest filesystem --format cyclonedx --output sbom.cdx.json /app`

And then open `sbom.cdx.json` and see the terraform dependencies. Currently, Trivy doesn't discover those dependencies so `sbom.cdx.json` doesn't list any dependencies.

### Target

Filesystem

### Scanner

Vulnerability

**Comment by simar7** [comment]
@candrews thanks for your question, to summarize you are looking at having an SBoM for IaC, in this case Terraform specifically? We've thought about it in the past but didn't get to a conclusion. @knqyf263 what do you think?

**Reply by candrews** [reply]
>  to summarize you are looking at having an SBoM for IaC, in this case Terraform specifically?

Yes, exactly :+1:

**Reply by knqyf263** [reply]
It sounds good, but I'm unsure if we can maintain it. Due to the lack of standardization, even regular SBOMs have compatibility problems, and we've been spending so much time on that. If something already has a strict specification and we just implement it accordingly, that might be fine.

**Comment by deejiw** [comment]
Hi all. I would like to share my practice that is currently under experimentation.
Instead of SBOM, I tried converting `.terraform.lock.hcl` to `json` using [`hcl2json` ](https://www.google.com/search?client=safari&rls=en&q=hcl+to+json&ie=UTF-8&oe=UTF-8) in order to perform direct interpolation, using `rego` custom policy, with command `trivy conf --misconfig-scanners json --config-check=policy --config-data=data` of which data directory contains `providers.yaml` that you might have a particular interest to assert/validate.

Sample of `providers.yaml`
```yaml
providers:
  registry.terraform.io/hashicorp/aws:
    min: 4.85.0
    max: 5.0.0
  registry.terraform.io/hashicorp/kubernetes:
    exact: 2.36.0
  registry.terraform.io/hashicorp/random:
    exact: 3.7.0
```
